developer/debugging/automated-tests.md

```diff
@@ -96,28 +109,46 @@ For instance, to run only the tests for the fedora-21 template, you can use the
     vm_qrexec_gui/TC_20_DispVM_fedora-21/test_010_simple_dvm_run
     vm_qrexec_gui/TC_20_DispVM_fedora-21/test_020_gui_app
     vm_qrexec_gui/TC_20_DispVM_fedora-21/test_030_edit_file
-    [user@dom0 ~]$ python3 -m qubes.tests.run -v `python3 -m qubes.tests.run -l | grep fedora-21`
+    [user@dom0 ~]$ sudo -E python3 -m qubes.tests.run -v `python3 -m qubes.tests.run -l | grep fedora-21`
 
 Example test run:
 
 ![snapshot-tests2.png](/attachment/wiki/developers/snapshot-tests2.png)
 
-### Qubes 4.0
+Tests are also compatible with nose2 test runner, so you can use this instead:
 
-Tests on Qubes 4.0 require stopping the `qubesd` service first, because a special instance of it is started as part of the test run.
-Additionally, tests needs to be started as root. The full command to run the tests is:
+    sudo systemctl stop qubesd; sudo -E nose2 -v --plugin nose2.plugins.loader.loadtests qubes.tests; sudo systemctl start qubesd
 
-    sudo systemctl stop qubesd; sudo -E python3 -m qubes.tests.run -v ; sudo systemctl start qubesd
+This may be especially useful together with various nose2 plugins to store tests results (for example `nose2.plugins.junitxml`), to ease presenting results. This is what we use on [OpenQA].
 
-On Qubes 4.0 tests are also compatible with nose2 test runner, so you can use this instead:
 
-    sudo systemctl stop qubesd; sudo -E nose2 -v --plugin nose2.plugins.loader.loadtests qubes.tests; sudo systemctl start qubesd
+### Unit testing inside a VM
 
-This may be especially useful together with various nose2 plugins to store tests results (for example `nose2.plugins.junitxml`), to ease presenting results. This is what we use on [OpenQA].
+Many unit tests will also work inside a VM. However all of the tests requiring a dedicated VM to be run (mostly the integration tests) will be skipped.
+
+As minimal requirements for testing, you'll have to clone the [qubes-core-admin](https://github.com/QubesOS/qubes-core-admin) and [qubes-core-qrexec](https://github.com/QubesOS/qubes-core-qrexec) repositories.
+
+The below example however will assume that you set up a build environment as described in the [Qubes Builder documentation](/doc/qubes-builder/).
+
+Assuming you cloned the `qubes-builder` repository to your home directory inside a fedora VM, you can use the following commands to run the unit tests:
+```{.bash}
+sudo dnf install python3-pip lvm2 python35 python3-virtualenv
+virtualenv -p /usr/bin/python35 python35
```

Comment:
This shouldn't be strictly necessary (code written for python 3.5 should work on newer python too), but indeed can improve test results (detect things that do work on python 3.7 for example, but will break on python 3.5).

But more importantly - the version to test on depends on Qubes version (what runs in dom0):
 - Qubes 4.0 (`release4.0` branch): python >=3.5
 - Qubes 4.1 (`master` branch): python >=3.7